File: algorithms/java/src/test/java/leetcode/LC345ReverseVowelsOfAString.java
package leetcode.lc345_reverse_vowels_of_a_string;

import org.junit.jupiter.api.Test;


import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/reverse-vowels-of-a-string/
 * #easy
 */
public final class LC345ReverseVowelsOfAString {
    public final class Solution {
        private boolean isVowel(final char letter) {
            char[] vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'};
            for (char vowel : vowels) {
                if (vowel == letter) {
                    return true;
                }
            }
            return false;
        }

        public String reverseVowels(final String s) {
            char[] text = s.toCharArray();
            for (int a = 0, z = text.length - 1; a < z; ) {
                while (a < text.length && !isVowel(text[a])) {
                    a++;
                }
                while (z >= 0 && !isVowel(text[z])) {
                    z--;
                }
                if (a < z) {
                    char exchange = text[a];
                    text[a] = text[z];
                    text[z] = exchange;
                    a++;
                    z--;
                }
            }
            return new String(text);
        }
    }

    @Test
    public void test_example_1() throws Exception {
        assertEquals("holle", new Solution().reverseVowels("hello"));
    }

    @Test
    public void test_example_2() throws Exception {
        assertEquals("leetcode", new Solution().reverseVowels("leotcede"));
    }

    @Test
    public void test_OE() throws Exception {
        assertEquals("EO", new Solution().reverseVowels("OE"));
    }

    @Test
    public void test_zt() throws Exception {
        assertEquals("zt", new Solution().reverseVowels("zt"));
    }
}

File: algorithms/code/leetcode/lc345_reverse_vowels_of_a_string/lc345_reverse_vowels_of_a_string.py
# https://leetcode.com/problems/reverse-vowels-of-a-string/

import unittest


class Solution:
    def __isVowel(self, letter: str) -> bool:
        vowels = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']
        return any(vowel == letter for vowel in vowels)

    def reverseVowels(self, s: str) -> str:
        text = list(s)
        a = 0
        z = len(text) - 1
        while a < z:
            while a < len(text) and not self.__isVowel(text[a]):
                a += 1
            while z >= 0 and not self.__isVowel(text[z]):
                z -= 1
            if a < z:
                exchange = text[a]
                text[a] = text[z]
                text[z] = exchange
                a += 1
                z -= 1
        return ''.join(text)


class TestCode(unittest.TestCase):
    def test_example_1(self) -> None:
        assert Solution().reverseVowels('hello') == 'holle'

    def test_example_2(self) -> None:
        assert Solution().reverseVowels('leotcede') == 'leetcode'

    def test_oe(self) -> None:
        assert Solution().reverseVowels('OE') == 'EO'

    def test_zt(self) -> None:
        assert Solution().reverseVowels('zt') == 'zt'
